File: SimpleMoxieSwitcher/Sources/SimpleMoxieSwitcher/Services/IntentDetectionService.swift
import Foundation

protocol IntentDetectionServiceProtocol {
    func detectIntent(from messages: [ChatMessage]) -> (intent: SessionIntent, confidence: Double)
    func detectDrift(currentIntent: SessionIntent, recentMessages: [ChatMessage]) -> Bool
    func generateRedirectionSuggestion(from current: SessionIntent, to detected: SessionIntent) -> String?
}

final class IntentDetectionService: IntentDetectionServiceProtocol {

    // MARK: - Intent Keywords

    private let playKeywords = [
        "play", "game", "fun", "silly", "joke", "laugh", "pretend", "imagine",
        "let's play", "wanna play", "can we play"
    ]

    private let learnKeywords = [
        "learn", "teach", "show me", "how to", "what is", "why", "explain",
        "help me understand", "I want to know", "can you teach", "homework",
        "study", "practice", "lesson"
    ]

    private let comfortKeywords = [
        "sad", "scared", "worried", "upset", "lonely", "miss", "hurt", "cry",
        "afraid", "nervous", "anxious", "feel bad", "not good", "help me feel",
        "I need", "comfort me"
    ]

    private let exploreKeywords = [
        "what if", "curious", "wonder", "explore", "discover", "find out",
        "show me around", "let's look", "I wonder", "tell me about"
    ]

    private let socializingKeywords = [
        "hi", "hello", "how are you", "what's up", "tell me about you",
        "let's talk", "chat", "friend", "buddy", "wanna hang out"
    ]

    private let storytellingKeywords = [
        "story", "tell me a story", "once upon", "adventure", "tale",
        "read me", "storytime", "bedtime story", "make up a story"
    ]

    // MARK: - Subject Detection for Learning

    private let subjects = [
        "math": ["math", "addition", "subtraction", "numbers", "counting", "multiply", "divide"],
        "science": ["science", "experiment", "animals", "plants", "space", "earth", "nature"],
        "reading": ["reading", "letters", "words", "alphabet", "spelling", "book"],
        "art": ["art", "drawing", "painting", "colors", "creative", "craft"],
        "social": ["feelings", "emotions", "friends", "sharing", "kindness", "manners"]
    ]

    // MARK: - Intent Detection

    func detectIntent(from messages: [ChatMessage]) -> (intent: SessionIntent, confidence: Double) {
        // Analyze recent messages (last 5 or all if less)
        let recentMessages = messages.suffix(5)
        let text = recentMessages
            .filter { $0.isUser }
            .map { $0.content.lowercased() }
            .joined(separator: " ")

        var scores: [SessionIntent: Double] = [
            .play: 0.0,
            .learn(subject: nil): 0.0,
            .comfort: 0.0,
            .explore: 0.0,
            .socializing: 0.0,
            .storytelling: 0.0
        ]

        // Score each intent
        scores[.play] = score(text: text, keywords: playKeywords)
        scores[.comfort] = score(text: text, keywords: comfortKeywords) * 1.2  // Weight comfort higher
        scores[.explore] = score(text: text, keywords: exploreKeywords)
        scores[.socializing] = score(text: text, keywords: socializingKeywords)
        scores[.storytelling] = score(text: text, keywords: storytellingKeywords)

        // Check for learning with subject detection
        let learnScore = score(text: text, keywords: learnKeywords)
        let subject = detectSubject(from: text)
        scores[.learn(subject: subject)] = learnScore

        // Detect questions (often learning or exploring)
        if text.contains("?") {
            if learnScore > 0 {
                scores[.learn(subject: subject)]? += 0.3
            } else {
                scores[.explore]? += 0.2
            }
        }

        // Find highest score
        let sorted = scores.sorted { $0.value > $1.value }
        guard let top = sorted.first else {
            return (.unknown, 0.0)
        }

        // If score is too low, return unknown
        if top.value < 0.3 {
            return (.unknown, top.value)
        }

        // Check if it's socializing by default (greetings at start)
        if messages.count <= 2 && scores[.socializing] ?? 0.0 > 0 {
            return (.socializing, scores[.socializing] ?? 0.0)
        }

        return (top.key, min(top.value, 1.0))
    }

    // MARK: - Drift Detection

    func detectDrift(currentIntent: SessionIntent, recentMessages: [ChatMessage]) -> Bool {
        let (detected, confidence) = detectIntent(from: recentMessages)

        // High confidence in a different intent suggests drift
        if detected != currentIntent && confidence > 0.6 {
            return true
        }

        return false
    }

    // MARK: - Redirection Suggestions

    func generateRedirectionSuggestion(from current: SessionIntent, to detected: SessionIntent) -> String? {
        switch (current, detected) {
        case (.socializing, .learn):
            return "I notice you're curious about learning something new! Want to start a lesson together?"

        case (.play, .learn):
            return "Sounds like you want to learn! Should we switch to learning mode?"

        case (.learn, .play):
            return "Ready for a break from learning? Let's have some fun!"

        case (.learn, .comfort):
            return "I can tell you might need a little break. Want to just talk for a bit?"

        case (.play, .comfort):
            return "Is everything okay? I'm here if you need to talk."

        case (_, .storytelling):
            return "Would you like me to tell you a story?"

        case (.socializing, .explore):
            return "You seem really curious! Want to explore and discover some new things together?"

        default:
            return nil
        }
    }

    // MARK: - Helper Methods

    private func score(text: String, keywords: [String]) -> Double {
        var score = 0.0
        let words = text.split(separator: " ").map { String($0) }

        for keyword in keywords {
            let keywordWords = keyword.split(separator: " ").map { String($0) }

            if keywordWords.count == 1 {
                // Single word keyword
                if words.contains(keyword) {
                    score += 0.3
                }
            } else {
                // Multi-word phrase
                if text.contains(keyword) {
                    score += 0.5  // Higher weight for exact phrases
                }
            }
        }

        return score
    }

    private func detectSubject(from text: String) -> String? {
        for (subject, keywords) in subjects {
            let subjectScore = score(text: text, keywords: keywords)
            if subjectScore > 0.3 {
                return subject
            }
        }
        return nil
    }
}

File: SimpleMoxieSwitcher-Windows/SimpleMoxieSwitcher/Services/IntentDetectionService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SimpleMoxieSwitcher.Models;

namespace SimpleMoxieSwitcher.Services
{
    /// <summary>
    /// Service for detecting user intent from conversation messages
    /// </summary>
    public class IntentDetectionService : IIntentDetectionService
    {
        // MARK: - Intent Keywords

        private readonly List<string> _playKeywords = new()
        {
            "play", "game", "fun", "silly", "joke", "laugh", "pretend", "imagine",
            "let's play", "wanna play", "can we play"
        };

        private readonly List<string> _learnKeywords = new()
        {
            "learn", "teach", "show me", "how to", "what is", "why", "explain",
            "help me understand", "I want to know", "can you teach", "homework",
            "study", "practice", "lesson"
        };

        private readonly List<string> _comfortKeywords = new()
        {
            "sad", "scared", "worried", "upset", "lonely", "miss", "hurt", "cry",
            "afraid", "nervous", "anxious", "feel bad", "not good", "help me feel",
            "I need", "comfort me"
        };

        private readonly List<string> _exploreKeywords = new()
        {
            "what if", "curious", "wonder", "explore", "discover", "find out",
            "show me around", "let's look", "I wonder", "tell me about"
        };

        private readonly List<string> _socializingKeywords = new()
        {
            "hi", "hello", "how are you", "what's up", "tell me about you",
            "let's talk", "chat", "friend", "buddy", "wanna hang out"
        };

        private readonly List<string> _storytellingKeywords = new()
        {
            "story", "tell me a story", "once upon", "adventure", "tale",
            "read me", "storytime", "bedtime story", "make up a story"
        };

        // MARK: - Subject Detection for Learning

        private readonly Dictionary<string, List<string>> _subjects = new()
        {
            ["math"] = new() { "math", "addition", "subtraction", "numbers", "counting", "multiply", "divide" },
            ["science"] = new() { "science", "experiment", "animals", "plants", "space", "earth", "nature" },
            ["reading"] = new() { "reading", "letters", "words", "alphabet", "spelling", "book" },
            ["art"] = new() { "art", "drawing", "painting", "colors", "creative", "craft" },
            ["social"] = new() { "feelings", "emotions", "friends", "sharing", "kindness", "manners" }
        };

        // MARK: - Intent Detection

        public (SessionIntent intent, double confidence) DetectIntent(List<ConversationMessage> messages)
        {
            // Analyze recent messages (last 5 or all if less)
            var recentMessages = messages.TakeLast(5).ToList();
            var text = string.Join(" ", recentMessages
                .Where(m => m.IsUser)
                .Select(m => m.Content.ToLower()));

            var scores = new Dictionary<string, double>
            {
                ["play"] = 0.0,
                ["learn"] = 0.0,
                ["comfort"] = 0.0,
                ["explore"] = 0.0,
                ["socializing"] = 0.0,
                ["storytelling"] = 0.0
            };

            // Score each intent
            scores["play"] = Score(text, _playKeywords);
            scores["comfort"] = Score(text, _comfortKeywords) * 1.2;  // Weight comfort higher
            scores["explore"] = Score(text, _exploreKeywords);
            scores["socializing"] = Score(text, _socializingKeywords);
            scores["storytelling"] = Score(text, _storytellingKeywords);

            // Check for learning with subject detection
            var learnScore = Score(text, _learnKeywords);
            var subject = DetectSubject(text);
            scores["learn"] = learnScore;

            // Detect questions (often learning or exploring)
            if (text.Contains("?"))
            {
                if (learnScore > 0)
                {
                    scores["learn"] += 0.3;
                }
                else
                {
                    scores["explore"] += 0.2;
                }
            }

            // Find highest score
            var sorted = scores.OrderByDescending(kv => kv.Value).ToList();
            var top = sorted.First();

            // If score is too low, return unknown
            if (top.Value < 0.3)
            {
                return (SessionIntent.Unknown, top.Value);
            }

            // Check if it's socializing by default (greetings at start)
            if (messages.Count <= 2 && scores["socializing"] > 0)
            {
                return (SessionIntent.Socializing, scores["socializing"]);
            }

            // Convert to SessionIntent enum
            var intent = top.Key switch
            {
                "play" => SessionIntent.Play,
                "learn" => subject != null ? SessionIntent.Learn : SessionIntent.Learn,
                "comfort" => SessionIntent.Comfort,
                "explore" => SessionIntent.Explore,
                "socializing" => SessionIntent.Socializing,
                "storytelling" => SessionIntent.Storytelling,
                _ => SessionIntent.Unknown
            };

            return (intent, Math.Min(top.Value, 1.0));
        }

        // MARK: - Drift Detection

        public bool DetectDrift(SessionIntent currentIntent, List<ConversationMessage> recentMessages)
        {
            var (detected, confidence) = DetectIntent(recentMessages);

            // High confidence in a different intent suggests drift
            if (detected != currentIntent && confidence > 0.6)
            {
                return true;
            }

            return false;
        }

        // MARK: - Redirection Suggestions

        public string GenerateRedirectionSuggestion(SessionIntent from, SessionIntent to)
        {
            return (from, to) switch
            {
                (SessionIntent.Socializing, SessionIntent.Learn) =>
                    "I notice you're curious about learning something new! Want to start a lesson together?",

                (SessionIntent.Play, SessionIntent.Learn) =>
                    "Sounds like you want to learn! Should we switch to learning mode?",

                (SessionIntent.Learn, SessionIntent.Play) =>
                    "Ready for a break from learning? Let's have some fun!",

                (SessionIntent.Learn, SessionIntent.Comfort) =>
                    "I can tell you might need a little break. Want to just talk for a bit?",

                (SessionIntent.Play, SessionIntent.Comfort) =>
                    "Is everything okay? I'm here if you need to talk.",

                (_, SessionIntent.Storytelling) =>
                    "Would you like me to tell you a story?",

                (SessionIntent.Socializing, SessionIntent.Explore) =>
                    "You seem really curious! Want to explore and discover some new things together?",

                _ => null
            };
        }

        // MARK: - Helper Methods

        private double Score(string text, List<string> keywords)
        {
            double score = 0.0;
            var words = text.Split(' ').ToList();

            foreach (var keyword in keywords)
            {
                var keywordWords = keyword.Split(' ').ToList();

                if (keywordWords.Count == 1)
                {
                    // Single word keyword
                    if (words.Contains(keyword))
                    {
                        score += 0.3;
                    }
                }
                else
                {
                    // Multi-word phrase
                    if (text.Contains(keyword))
                    {
                        score += 0.5;  // Higher weight for exact phrases
                    }
                }
            }

            return score;
        }

        private string DetectSubject(string text)
        {
            foreach (var (subject, keywords) in _subjects)
            {
                var subjectScore = Score(text, keywords);
                if (subjectScore > 0.3)
                {
                    return subject;
                }
            }
            return null;
        }
    }

    // MARK: - Interface

    public interface IIntentDetectionService
    {
        (SessionIntent intent, double confidence) DetectIntent(List<ConversationMessage> messages);
        bool DetectDrift(SessionIntent currentIntent, List<ConversationMessage> recentMessages);
        string GenerateRedirectionSuggestion(SessionIntent from, SessionIntent to);
    }

    // MARK: - Session Intent Enum

    public enum SessionIntent
    {
        Unknown,
        Play,
        Learn,
        Comfort,
        Explore,
        Socializing,
        Storytelling
    }
}
